fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java

```diff
@@ -0,0 +1,371 @@
+            return false;
+        }
+        return apply.left() != null && apply.right() instanceof LogicalAggregate
+                && ((LogicalAggregate<?>) apply.right()).child() != null;
+    }
+
+    private boolean check(LogicalFilter<LogicalProject<LogicalApply<Plan, LogicalAggregate<Plan>>>> node) {
+        LogicalApply<?, ?> apply = node.child().child();
+        LogicalPlan outer = ((LogicalPlan) apply.child(0));
+        LogicalPlan inner = ((LogicalPlan) apply.child(1));
+        outerPlans = new PlanCollector().collect(outer);
+        innerPlans = new PlanCollector().collect(inner);
+        LogicalFilter outerFilter = node;
+        Optional<LogicalFilter> innerFilter = innerPlans.stream()
+                .filter(LogicalFilter.class::isInstance)
+                .map(LogicalFilter.class::cast).findFirst();
+        return innerFilter.filter(
+                logicalFilter -> checkPlanType() && checkAggType()
+                        && checkRelation(apply.getCorrelationSlot())
+                        && checkPredicate(Sets.newHashSet(outerFilter.getConjuncts()),
+                        Sets.newHashSet(logicalFilter.getConjuncts()))
+        ).isPresent();
+    }
+
+    private Plan trans(LogicalFilter<LogicalProject<LogicalApply<Plan, LogicalAggregate<Plan>>>> node) {
+        LogicalApply<Plan, LogicalAggregate<Plan>> apply = node.child().child();
+        LogicalAggregate<Plan> agg = apply.right();
+
+        // for example, in tpc-h Q17
+        // window filter conjuncts is cast(l_quantity#id1 as decimal(27, 9)) < `0.2 * avg(l_quantity)`#id2 and
+        // 0.2 * avg(l_quantity#id3) as `0.2 * l_quantity`#id2 of agg's output expression
+        // we change it to cast(l_quantity#id1 as decimal(27, 9)) < 0.2 * `avg(l_quantity#id1) over(window)`#id4
+        // and avg(l_quantity#id1) over(window) as `avg(l_quantity#id1) over(window)`#id4
+        // first we get the slots of the two sides of apply
+
+        Preconditions.checkArgument(apply.getSubCorrespondingConject().get() instanceof ComparisonPredicate);
+        ComparisonPredicate windowFilterConjunct = PlanUtils.maybeCommuteComparisonPredicate(
+                ((ComparisonPredicate) apply.getSubCorrespondingConject().get()), apply.left());
+
+        // build window function, replace the slot
+        List<Expression> windowAggSlots = windowFilterConjunct.left().collectToList(Slot.class::isInstance);
+        Preconditions.checkArgument(windowAggSlots.size() == 1);
+
+        // adjust agg function's nullable.
+        AggregateFunction function = functions.get(0);
+        if (function instanceof NullableAggregateFunction) {
+            function = ((NullableAggregateFunction) function).withAlwaysNullable(false);
+        }
+
+        WindowExpression windowFunction = createWindowFunction(apply.getCorrelationSlot(),
+                function.withChildren(windowAggSlots));
+        NamedExpression windowFunctionAlias = new Alias(windowFunction, "wf");
```

Comment:
u need use a unique name generate by column name generator
refer to `statementContext.generateColumnName()`